File: data_structures/linked_list/doubly_linked_list.dart
class Node {
  dynamic data;
  Node? prev;
  Node? next;

  Node(this.data);
}

class DoublyLinkedList {
  Node? head;
  Node? tail;

  // Insertion at the beginning of the list
  void insertAtBeginning(dynamic data) {
    Node newNode = Node(data);
    if (head == null) {
      head = newNode;
      tail = newNode;
    } else {
      newNode.next = head;
      head!.prev = newNode;
      head = newNode;
    }
  }

  // Insertion at the end of the list
  void insertAtEnd(dynamic data) {
    Node newNode = Node(data);
    if (head == null) {
      head = newNode;
      tail = newNode;
    } else {
      newNode.prev = tail;
      tail!.next = newNode;
      tail = newNode;
    }
  }

  // Insertion at a specific position
  void insertAtPosition(dynamic data, int position) {
    Node newNode = Node(data);
    if (position < 0) position = 0;

    if (head == null || position == 0) {
      insertAtBeginning(data);
    } else {
      Node? current = head;
      for (int i = 0; i < position - 1 && current != null; i++) {
        current = current.next;
      }

      if (current == null) {
        insertAtEnd(data);
      } else {
        newNode.next = current.next;
        newNode.prev = current;
        current.next!.prev = newNode;
        current.next = newNode;
      }
    }
  }

  // Deletion from the beginning of the list
  void deleteFromBeginning() {
    if (head == null) return;

    if (head == tail) {
      head = null;
      tail = null;
    } else {
      head = head!.next;
      head!.prev = null;
    }
  }

  // Deletion from the end of the list
  void deleteFromEnd() {
    if (tail == null) return;

    if (head == tail) {
      head = null;
      tail = null;
    } else {
      tail = tail!.prev;
      tail!.next = null;
    }
  }

  // Deletion from a specific position
  void deleteFromPosition(int position) {
    if (head == null || position < 0) return;

    if (position == 0) {
      deleteFromBeginning();
    } else {
      Node? current = head;
      for (int i = 0; i < position && current != null; i++) {
        current = current.next;
      }

      if (current == null) return;

      if (current == tail) {
        deleteFromEnd();
      } else {
        current.prev!.next = current.next;
        current.next!.prev = current.prev;
      }
    }
  }

  // Display the elements of the list
  void display() {
    Node? current = head;
    while (current != null) {
      print(current.data);
      current = current.next;
    }
  }
}

void main() {
  DoublyLinkedList myList = DoublyLinkedList();

  myList.insertAtEnd(1);
  myList.insertAtEnd(2);
  myList.insertAtEnd(3);

  print("Original List:");
  myList.display();

  myList.insertAtBeginning(0);
  myList.insertAtEnd(4);
  myList.insertAtPosition(2.5, 3);

  print("\nList after Insertions:");
  myList.display();

  myList.deleteFromBeginning();
  myList.deleteFromEnd();
  myList.deleteFromPosition(2);

  print("\nList after Deletions:");
  myList.display();
}

File: go/data_structures/linked_list/doubly_linked_list.go
// doubly_linked_list.go

package main

import "fmt"
type Node struct {
    data int
    prev *Node
    next *Node
}
type DoublyLinkedList struct {
    head *Node
    tail *Node
}

func (dll *DoublyLinkedList) InsertAtBeginning(data int) {
    newNode := &Node{data: data}
    if dll.head == nil {
        dll.head = newNode
        dll.tail = newNode
        return
    }

    newNode.next = dll.head
    dll.head.prev = newNode
    dll.head = newNode
}

func (dll *DoublyLinkedList) InsertAtEnd(data int) {
    newNode := &Node{data: data}
    if dll.head == nil {
        dll.head = newNode
        dll.tail = newNode
        return
    }

    newNode.prev = dll.tail
    dll.tail.next = newNode
    dll.tail = newNode
}

func (dll *DoublyLinkedList) InsertAtPosition(data, position int) {
    newNode := &Node{data: data}
    if dll.head == nil || position == 0 {
        dll.InsertAtBeginning(data)
        return
    }

    currentNode := dll.head
    for i := 0; i < position-1 && currentNode != nil; i++ {
        currentNode = currentNode.next
    }

    if currentNode == nil {
        return
    }

    newNode.prev = currentNode
    newNode.next = currentNode.next
    currentNode.next.prev = newNode
    currentNode.next = newNode
}

func (dll *DoublyLinkedList) DeleteFromBeginning() {
    if dll.head == nil {
        return
    }

    if dll.head == dll.tail {
        dll.head = nil
        dll.tail = nil
        return
    }

    dll.head = dll.head.next
    dll.head.prev = nil
}

func (dll *DoublyLinkedList) DeleteFromEnd() {
    if dll.head == nil {
        return
    }

    if dll.head == dll.tail {
        dll.head = nil
        dll.tail = nil
        return
    }

    dll.tail = dll.tail.prev
    dll.tail.next = nil
}

func (dll *DoublyLinkedList) DeleteFromPosition(position int) {
    if dll.head == nil {
        return
    }

    if position == 0 {
        dll.DeleteFromBeginning()
        return
    }

    currentNode := dll.head
    for i := 0; i < position && currentNode != nil; i++ {
        currentNode = currentNode.next
    }

    if currentNode == nil {
        return
    }

    if currentNode == dll.tail {
        dll.tail = currentNode.prev
        dll.tail.next = nil
    } else {
        currentNode.prev.next = currentNode.next
        currentNode.next.prev = currentNode.prev
    }
}

func (dll *DoublyLinkedList) DisplayForward() {
    currentNode := dll.head
    for currentNode != nil {
        fmt.Printf("%d <-> ", currentNode.data)
        currentNode = currentNode.next
    }
    fmt.Println("nil")
}

func (dll *DoublyLinkedList) DisplayBackward() {
    currentNode := dll.tail
    for currentNode != nil {
        fmt.Printf("%d <-> ", currentNode.data)
        currentNode = currentNode.prev
    }
    fmt.Println("nil")
}

func main() {
    doublyLinkedList := &DoublyLinkedList{}

    doublyLinkedList.InsertAtEnd(10)
    doublyLinkedList.InsertAtBeginning(5)
    doublyLinkedList.InsertAtEnd(20)
    doublyLinkedList.InsertAtPosition(15, 2)

    fmt.Println("Original Doubly Linked List (Forward Order):")
    doublyLinkedList.DisplayForward()

    fmt.Println("Original Doubly Linked List (Backward Order):")
    doublyLinkedList.DisplayBackward()

    doublyLinkedList.DeleteFromBeginning()
    doublyLinkedList.DeleteFromEnd()
    doublyLinkedList.DeleteFromPosition(1)

    fmt.Println("Updated Doubly Linked List (Forward Order):")
    doublyLinkedList.DisplayForward()

    fmt.Println("Updated Doubly Linked List (Backward Order):")
    doublyLinkedList.DisplayBackward()
}
